docs/specs/3-13-2023-commands-fit-n-finsh.md

```diff
@@ -0,0 +1,71 @@
+
+```
+Commands
+  Develop an app
+    init     	: Initialize a new application.
+    provision	: Provision the Azure resources for an application.
+    deploy   	: Deploy the application's code to Azure.
+    up       	: Initialize application, provision Azure resources, and deploy your project with a single command.
+
+  Monitor, test, and release your app
+    monitor  	: Monitor a deployed application.
+    pipeline 	: Manage and configure your deployment pipelines.
+
+  Authenticate with Azure
+    login    	: Log in to Azure.
+    logout   	: Log out of Azure.
+
+  Miscellaneous commands
+    config   	: Manage azd configurations (ex: default Azure subscription, location).
+    down     	: Delete Azure resources for an application.
+    env      	: Manage environments.
+    restore  	: Restore application dependencies.
+    template 	: Find and view template details.
+```
+
+The commands are grouped by their scenarios, with most important scenarios at the top. It is not in alphabetical order, rather the order the user is expected to run the commands. If we added `package` for example, would be after `init`.
+
+## Command changes
+
+The following changes are proposed:
+
+- `azd infra create` and `azd infra delete` should be removed, as the functionality already exists in `azd provision` and `azd down`, and users will find the duplication somewhat confusing.
+- The `--output` flag should be moved to `experimental` (hidden and unsupported unless with feature toggle), with the exception of `azd auth token --output json` which must be supported due to credential handlers.
+  - When the feature is hidden, the default output behavior of each command will apply.
+  - This is done as `--output` does not have a documented contract. It prints out `stdout` in JSON, while `stderr` messages also contain JSON-enveloped messages. We might consider splitting this behavior in the future.
+
+Discussion warranted:
+
+- Should we offer `azd restore`? I'm curious if users actually find this useful. We currently use this in our `tasks.json`, which seems useful, but the immediate block directly references `npm`, which defeats the nice abstraction. Excerpt from `tasks.json` below:
+  ```json
+    {
+    "label": "Restore Web",
+    "type": "shell",
+    "command": "azd restore --service web",
+    "presentation": {
+        "reveal": "silent"
+    },
+    "problemMatcher": []
+    },
+    {
+        "label": "Web npm start",
+        "detail": "Helper task--use 'Start Web' task to ensure environment is set up correctly",
+        "type": "shell",
+        "command": "npm run start"
+    }
+  ```
+- Should `azd deploy` deploy only the current service, when the working directory is under a service directory? I personally think this is what users might expect, but I can see an argument against it since it may be implicit behavior.
```

Comment:
I like it how it is - `azd deploy` deploys everything, `azd deploy --service` deploys a particular service.   I don't want to have to go into each service and call deploy for each one when I want to deploy the whole app.

Reply:
I am not sure thats the question. I think the question is, if the CWD is the service, should azd default to just deploying that service which means probing for the azure.yaml and then running with the --service flag from the cwd. Safest choice here if we don't have a design is to fail these commands if they aren't run with the CWD having the azure.yaml m, no?

Reply:
That was the original proposal was to make azd deploy not deploy the whole solution....

Reply:
They're orthogonal proposals to one another. This one specifically changes only the behavior when CWD is inside a service directory. I've separated them out.

Reply:
How do our azd commands generally behave when they are run from subfolders?

Reply:
> How do our azd commands generally behave when they are run from subfolders?

In general, as long as there is an `azure.yaml` file in your current working directory, or some folder above, the gestures "just work".  We don't use your location in in the tree rooted by `azure.yaml` today to inform how any commands run, so in practice they all should behave as if you ran them from the root of your project.

Reply:
So if we don't find an azure.yaml in the cwd then we walk up until we find one?

Reply:
Looks like we do walk up until we find one. Else error out. 

If in future we want to make `azd deploy` contextual. Then we should not do the look up and just error out now?. Keeping a placeholder for us to add `azd deploy` which will run just the contextual service.

Reply:
I've captured all the thoughts and suggestions in `#1491` in our private repo. Everyone is tagged in it.